### PR TITLE
bpo-34041: Added *deterministic* parameter to sqlite3.Connection.create_function().

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://bugs.python.org/issue34041

### CODE REVIEW

File: Modules/_sqlite/connection.c
Comment (vstinner) [inline Modules/_sqlite/connection.c]: Why do you have to define the constant in Python? Does it mean that SQLite doesn't support the flag? I would prefer to raise an exception, rather than using an hardcoded constant.

File: Modules/_sqlite/connection.c
Comment (vstinner) [inline Modules/_sqlite/connection.c]: Maybe it would be safer to use a keyword only parameter, to prevent bugs?

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Modules/_sqlite/connection.c
Comment (sir-sigurd) [reply Modules/_sqlite/connection.c]: Do you think an exception should raised if Python is compiled with SQLite version that doesn't support this feature but is runned with SQLite version that supports it?

Comment (vstinner) [reply Modules/_sqlite/connection.c]: What happens if you pass 0x800 to a SQLite which doesn't know/support this flag? I expect an error.

Comment (sir-sigurd) [reply Modules/_sqlite/connection.c]: I checked it with SQLite 3.8.2 (the most ancient dockerized version I found). Function is created, but it's not marked as deterministic, so optimizations are not used, it's reflected in the added test.

Comment (sir-sigurd) [reply Modules/_sqlite/connection.c]: IMHO it doesn't make sense to raise exception here because if `deterministic` is requested, but created function is not marked as deterministic this only affects performance, but indeed, flag should not be passed if it's unsupported as it was working properly only by accident.

File: Modules/_sqlite/connection.c
Comment (vstinner) [inline Modules/_sqlite/connection.c]: Hum, if the flag only has an impact on performance, IHMO it's ok to ignore it if SQLite doesn't support it. But I would prefer to use "#ifndef SQLITE_DETERMINISTIC" for this test. Something like:

#ifndef SQLITE_DETERMINISTIC
if (deterministic) eTextRep |= SQLITE_DETERMINISTIC;
#endif

If Python is built on old SQLite: no luck, you will never get deterministic functions. If you build Python on new SQLite but run old SQLite: SQLITE_DETERMINISTIC is ignored, it's fine. If you have new versions in both cases: you're lucky, and you get the optimization :-)

File: Doc/library/sqlite3.rst
Comment (serhiy-storchaka) [inline Doc/library/sqlite3.rst]: I would make the new boolean parameter a keyword-only.

Comment (vstinner) [reply Doc/library/sqlite3.rst]: Yep, I made the same proposal :-) At the C level, it means using "$" instead of "|" before the parameter. Parameters after "$" are keyword-only parameters.

File: Modules/_sqlite/connection.c
Comment (sir-sigurd) [reply Modules/_sqlite/connection.c]: I carefully reread SQLite docs and it appears that non-deterministic functions can't be used in indexes, so it seems that exception should be raised after all.

**Comment by sir-sigurd** [comment]
I have made the requested changes; please review again

File: Doc/library/sqlite3.rst
Comment (berkerpeksag) [inline Doc/library/sqlite3.rst]: Nit: Indentation is off here:

```rst
.. versionchanged:: 3.8
   The *deterministic* parameter was added.
```

File: Misc/NEWS.d/next/Library/2018-07-04-21-24-58.bpo-34041.aKk8t5.rst
Comment (berkerpeksag) [inline Misc/NEWS.d/next/Library/2018-07-04-21-24-58.bpo-34041.aKk8t5.rst]: It would be better if we can link to the `create_function()` documentation:

```rst
.. untested
:meth:`sqlite3.Connection.create_function`
```

Comment (berkerpeksag) [reply Misc/NEWS.d/next/Library/2018-07-04-21-24-58.bpo-34041.aKk8t5.rst]: It seems like we are missing "method" here:

the sqlite3.Connection.create_function() -> the sqlite3.Connection.create_function() method

File: Doc/library/sqlite3.rst
Comment (berkerpeksag) [inline Doc/library/sqlite3.rst]: Could you please mention that `NotSupportedError` will be raised as well?

File: Modules/_sqlite/connection.c
Comment (berkerpeksag) [inline Modules/_sqlite/connection.c]: Style nit: We usually don't use camelCase convention in Python codebase.

File: Lib/sqlite3/test/userfunctions.py
Comment (berkerpeksag) [inline Lib/sqlite3/test/userfunctions.py:295]: Could you also add a simple test to make sure that `deterministic` is a keyword-only argument?

```py
self.con.create_function("deterministic", 0, int, True)
```

File: Misc/NEWS.d/next/Library/2018-07-04-21-24-58.bpo-34041.aKk8t5.rst
Comment (berkerpeksag) [reply Misc/NEWS.d/next/Library/2018-07-04-21-24-58.bpo-34041.aKk8t5.rst]: And please add "Patch by Sergey Fedoseev.".

File: Modules/_sqlite/connection.c
Comment (berkerpeksag) [inline Modules/_sqlite/connection.c]: We usually don't check for runtime library version when doing feature detection, but I think this is better. (Note that the only other check was added only in 2016: https://github.com/python/cpython/commit/7bea2347c7e8af2c6b59e541c4039c34c7a8f6b2)

**Review by berkerpeksag — CHANGES_REQUESTED** [review]

File: Lib/sqlite3/test/userfunctions.py
Comment (vstinner) [reply Lib/sqlite3/test/userfunctions.py:295]: fixed in the following test

File: Modules/_sqlite/connection.c
Comment (vstinner) [inline Modules/_sqlite/connection.c]: would you mind to add "deterministic"? "deterministic requires SQLite 3.8.3 or higher"

File: Modules/_sqlite/connection.c
Comment (vstinner) [inline Modules/_sqlite/connection.c]: I suggest to reuse the same error message: "deterministic requires SQLite 3.8.3 or higher"

File: Modules/_sqlite/connection.c
Comment (vstinner) [inline Modules/_sqlite/connection.c]: "else {" can be removed, since the if block returns.

File: Modules/_sqlite/connection.c
Comment (vstinner) [reply Modules/_sqlite/connection.c]: +1 for the runtime check

Checking the runtime version is fine. SQLITE_VERSION_NUMBER is the build version. Using Linux packages, it's common that the machine building packages and the users installing packages use different versions.

It also seems risky to use the flag on old versions according to the discussion in this PR.

**Review by vstinner — COMMENTED** [review]
Almost good! I just have a minor request on the error message.

File: Modules/_sqlite/connection.c
Comment (sir-sigurd) [reply Modules/_sqlite/connection.c]: maybe even "deterministic=True requires SQLite 3.8.3 or higher"?

Comment (vstinner) [reply Modules/_sqlite/connection.c]: yeah, =True is maybe even better (more explicit).

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
@serhiy-storchaka, @berkerpeksag: The change LGTM. Would you mind to also review the latest version?

File: Doc/library/sqlite3.rst
Comment (berkerpeksag) [inline Doc/library/sqlite3.rst:349]: I was going to suggest add a link to the `NotSupportedError` documentation, then I noticed it wasn't documented. I've opened https://bugs.python.org/issue34061

**Review by berkerpeksag — APPROVED** [review]
LGTM too. I will wait for a day or two to let Serhiy have a chance to review the latest revision, and merge it.

Thanks!

File: Lib/sqlite3/test/userfunctions.py
Comment (serhiy-storchaka) [inline Lib/sqlite3/test/userfunctions.py]: Why a mock is used?

**Review by serhiy-storchaka — APPROVED** [review]
LGTM if LGT @berkerpeksag.

I just do not fully understand the test.

File: Lib/sqlite3/test/userfunctions.py
Comment (sir-sigurd) [reply Lib/sqlite3/test/userfunctions.py]: It's used to test how much times function is called. If function is non-deterministic, query can't be optimized, so function have to be called 2 times (and actually there is no need to call it at all for this query :-).
This test relies on SQLite's behavior in some degree, but I did't find out a better way to test it.

Comment (serhiy-storchaka) [reply Lib/sqlite3/test/userfunctions.py]: Thank you for clarification.

Is it worth to add a test for `deterministic=False` (or for not specified)?

Comment (berkerpeksag) [reply Lib/sqlite3/test/userfunctions.py]: Perhaps we can do something like the following instead of using mock:

```py
def foo():
    foo.calls.append(True)
foo.calls = []
```

Then we can check the call count as:

```py
self.assertEqual(len(foo.calls), 1)
```

Comment (sir-sigurd) [reply Lib/sqlite3/test/userfunctions.py]: @serhiy-storchaka added test for it.
@berkerpeksag not sure why this could be better.

File: Lib/sqlite3/test/userfunctions.py
Comment (serhiy-storchaka) [inline Lib/sqlite3/test/userfunctions.py:295]: `self.con.create_function("deterministic", 0, int, deterministic=False)` should work even on SQLite < 3.8.3, right?

File: Lib/sqlite3/test/userfunctions.py
Comment (serhiy-storchaka) [reply Lib/sqlite3/test/userfunctions.py]: @berkerpeksag This is how I would write this test. But I just don't use mocks much. I think using a mock here is good.

File: Lib/sqlite3/test/userfunctions.py
Comment (sir-sigurd) [reply Lib/sqlite3/test/userfunctions.py:295]: Yes, just pushed the fix.

File: Lib/sqlite3/test/userfunctions.py
Comment (berkerpeksag) [inline Lib/sqlite3/test/userfunctions.py]: I think this test can be split into three tests:

1. `deterministic=False`
2. `deterministic=True` if SQLite is older than 3.8.3
3. `deterministic=True` if SQLite supports the feature

File: Lib/sqlite3/test/userfunctions.py
Comment (berkerpeksag) [inline Lib/sqlite3/test/userfunctions.py]: Nit: Kw -> Keyword

**Comment by berkerpeksag** [comment]
Thanks!

**Pull request merged**
